File: 1122Programming/leetcode/2485. Find the Pivot Integer/ans.rs
/**
 * https://leetcode.com/problems/find-the-pivot-integer/
 * Problem ID: 2485. Find the Pivot Integer
 *
 * Author: revival0728
 */


impl Solution {
    pub fn pivot_integer(n: i32) -> i32 {
        let n = n as usize;
        let mut prefix_sum = vec![0_usize];
        for i in 1..(n + 1) as usize {
            prefix_sum.push(i + prefix_sum.last().unwrap());
        }
        for x in 1..(n + 1) as usize {
            if prefix_sum[x] == prefix_sum[n] - prefix_sum[x - 1] {
                return x as i32;
            }
        }
        -1
    }
}

File: 1122Programming/leetcode/2485. Find the Pivot Integer/ans.cpp
/**
 * https://leetcode.com/problems/find-the-pivot-integer/
 * Problem ID: 2485. Find the Pivot Integer
 *
 * Author: Ateto
 */

class Solution {
private: 
	int cal(int p, int q) {
        int tmp = (p + q) * (q - p + 1) / 2;
        return tmp;
    }
public: 
    int pivotInteger(int n) {
        int left = 1;
        int right = n;
        while (left <= right) {
            int mid = (left + right) >> 1;
            int x = cal(1, mid);
            int y = cal(mid, n);
            if (x == y) return mid;
            if (x < y) {
                left = mid+1;
            } else {
                right = mid-1;
            }
        }
        return -1;
    }
};